### PR TITLE
[GPU] Fix issues in dynamic loops newly created by TF FE support

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
 - Implementation of loop for new TF FE(PR#20270)
 - Apply infinite trip count(-1)
 - Apply _exec_deps for backedge
 - Apply cached constant primitive id for loop construnction

### Tickets:
 - 123934
 - 123651

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/loop.cpp [hunk 1]
```diff
@@ -540,6 +540,23 @@ void loop_inst::preprocess_backedge_memory() {
...
+                user_primitive_id = user->get_primitive()->id;
+            }
+            auto user_primitive = body_network->get_primitive(user_primitive_id);
+
+            if (std::find(backedge_from_prim->_exec_deps.begin(), backedge_from_prim->_exec_deps.end(), user_primitive)
+               != backedge_from_prim->_exec_deps.end()) {
+                backedge_from_prim->_exec_dep_ids.push_back(user_primitive_id);
+                backedge_from_prim->_exec_deps.push_back(user_primitive);
+               }
+        }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/loop.cpp:559]: There might be a case where deadlock can happens. e.g., 
```
input(backedge_to) ------> op1 ----->
                        />
                     op2 ----->result (backedge_from)
```
                              
  Here if result node is opt out, then it will be  either of the following case:
  1)  op2 becomes the new backedge_from
  => op2 and op1 will have cyclic dependency 
  2) or, still the result is backedge but using op2's memory
  => above change won't be applied for op2 => original issue still happen 
  
So I think we should do additional thing for this purpose, i.e., we need to prevent network output node (reorder) 
 from being optimized out, if the output is in the loop body network && it is the backedge of mem sharing

File: src/plugins/intel_gpu/src/graph/loop.cpp [hunk 2]
```diff
@@ -540,6 +540,23 @@ void loop_inst::preprocess_backedge_memory() {
             GPU_DEBUG_LOG << idx << ") add back_edge mapping with SINGLE_SHARED type, backedge_mem("
                             << backedge_mem << "), initial_mem(" << initial_mem << ")" << std::endl;
         }
+        // Add _exec_deps for backedge primitives
+        for (auto user : backedge_to_prim->get_node().get_users()) {
+            cldnn::primitive_id user_primitive_id;
+            if (user->can_be_optimized()) {
+                // When user is optimized out, use user's user.
+                user_primitive_id = user->get_users().front()->get_primitive()->id;
```
[the remark this answers is not in the document]
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/loop.cpp]: UPdated!

File: src/plugins/intel_gpu/src/graph/loop.cpp (hunk 1 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/loop.cpp:566]: Updated the condition in remove_redundant_reorder pass. Thanks!

**Review by ahnyoung-paul — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp [hunk 3]
```diff
@@ -133,8 +133,13 @@ void remove_redundant_reorders::run(program& p) {
         auto node = *itr++;
         if (!node->is_type<reorder>())  // only care for reorders
             continue;
-
         auto& r_node = node->as<reorder>();
+
+        // Do not opt out result reorder of Loop body network
+        bool is_loop_body_network_output = (r_node.get_program().is_body_program && r_node.is_output() && r_node.can_share_buffer());
+        if (is_loop_body_network_output)
+            continue;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:141]: I think we should add addiitional condition as I mentioned before, 
1. is loob body output 
2. is backedge (from) and the the target node (to) is sharing memory

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:141]: Also, can_share_buffer is not relevant to the expected change.

File: src/plugins/intel_gpu/src/graph/loop.cpp [hunk 4]
```diff
@@ -540,6 +540,30 @@ void loop_inst::preprocess_backedge_memory() {
...
         }
+
+        if (backedge_to_prim->_node != nullptr) {
+            update_backedge_exec_deps(backedge_to_prim->get_node(), backedge_from_prim->id());
+        }
+    }
+}
+
+void loop_inst::update_backedge_exec_deps(const cldnn::program_node& node, const cldnn::primitive_id& backedge_from_prim_id) {
+    // Add _exec_deps for backedge primitives
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/loop.cpp:551]: Also could you please add a comment why this change is added?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/loop.cpp]: Updated! Thanks!

**Review by yeonbok — APPROVED** [review]
